### PR TITLE
Add phone and email link block

### PR DESCRIPTION
**Pull request by SebiVPS**
---

## PR Checklist

-   [x] Verify if the change requires a changeset. See [CONTRIBUTING.md](https://github.com/vivid-planet/comet/blob/HEAD/CONTRIBUTING.md)
-   [x] Provide screenshots/screencasts if the change contains visual changes

<details>
    <summary>Screenshots/screencasts</summary>
    
Screenshot 2024-06-05 at 08 20 08

</details>

### CODE REVIEW

**Comment by nsams** [comment]
I remember we discussed if we need this or not, but that is years ago. We decided No back then.

**Comment by SebiVPS** [comment]
@nsams: @johnnyomair and @thomasdax98 decided that we should add these two blocks to the library. Meanwhile there are multiple implementations in many different repos around. So instead of add another copy to the comet-starter, we add it to the library and remove the other copies. The components itself are pretty common and can be used for every use case when they are needed.

**Comment by johnnyomair** [comment]
@nsams the problem with supporting `mailto:` and `tel:` in ExternalLinkBlock only is that editors don't know about them. We thought about adding a fancy UI/UX feature to fix this (e.g., detecting emails automatically and add the mailto: prefix), but couldn't come up with an elegant solution. Furthermore, editors would still need to know that they have to select "External link" in the link block for a phone number, which isn't intuitive. So we decided having two more options is the way to go.

**Comment by johnnyomair** [comment]
@SebiVPS please rebase and resolve conflicts.

File: .changeset/fifty-lemons-work.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+"@comet/cms-admin": minor
+"@comet/blocks-api": minor
+"@comet/cms-site": minor
+"@comet/cms-api": minor
+---
+
+Add PhoneLinkBlock and EmailLinkBlock
```
Comment (johnnyomair) [inline .changeset/fifty-lemons-work.md:8]:
```suggestion
Add `PhoneLinkBlock` and `EmailLinkBlock`
```

File: packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import * as React from "react";
+import { FormattedMessage } from "react-intl";
+
+interface EmailLinkBlockData {
+    email?: string;
+}
+
+interface EmailLinkBlockInput {
+    email?: string;
+}
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx:12]: Please use the generated types here

File: packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        return {
+            email: state.email,
+        };
+    },
+
+    output2State: async (output) => {
+        return {
+            email: output.email,
+        };
+    },
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx:39]: I don't think it's necessary to add these methods, that's what createBlockSkeleton does

File: packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+};
+
+const isEmail = (text: string) => {
+    return !!String(text)
+        .toLowerCase()
+        .match(
+            /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
+        );
+};
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx:80]: We have class-validator as a dependency, please use their isEmail validator.

File: packages/admin/cms-admin/src/blocks/PhoneLinkBlock.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        );
+    },
+    previewContent: (state) => {
+        return state.phone ? [{ type: "text", content: state.phone }] : [];
+    },
+};
+
+const isPhone = (text: string) => {
+    return !!String(text).match(/^[+]?[0-9]{4,20}$/im);
+};
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/blocks/PhoneLinkBlock.tsx:76]: We already have a phone number validation in [isLinkTarget](https://github.com/vivid-planet/comet/blob/apigen-enumsarray-filter/packages/admin/cms-admin/src/validation/isLinkTarget.ts#L13). Please create a function that can be used by the block and the validator

File: packages/api/blocks-api/src/blocks/phone-link.block.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { IsString } from "class-validator";
+
+import { BlockData, BlockInput, createBlock, inputToData } from "./block";
+import { BlockField } from "./decorators/field";
+
+class PhoneLinkBlockData extends BlockData {
+    @BlockField()
```
Comment (johnnyomair) [inline packages/api/blocks-api/src/blocks/phone-link.block.ts:7]: Please add nullable here

File: packages/api/blocks-api/src/blocks/phone-link.block.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    phone: string;
+}
+
+class PhoneLinkBlockInput extends BlockInput {
+    @IsString()
+    @BlockField()
+    phone: string;
```
Comment (johnnyomair) [inline packages/api/blocks-api/src/blocks/phone-link.block.ts:14]: Please add the phone number validation from [IsLinkTarget](https://github.com/vivid-planet/comet/blob/main/packages/api/blocks-api/src/blocks/externalLinkBlock/is-link-target.validator.ts#L32) here

File: packages/api/blocks-api/src/blocks/validator/is-phone-number.validator.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import { Injectable } from "@nestjs/common";
+import { isString, registerDecorator, ValidatorConstraint, ValidatorConstraintInterface } from "class-validator";
+
+const PHONE_NUMBER_REGEX = /^\+?[0-9\s]+$/;
+
+export const testPhoneNumber = (value: string): boolean => {
```
[the remark this answers is not in the document]
Comment (johnnyomair) [reply packages/api/blocks-api/src/blocks/validator/is-phone-number.validator.ts:6]: You could just use this name 😁 class-validator also has some exports both lower and upper case.

File: packages/admin/cms-admin/src/blocks/PhoneLinkBlock.tsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                        updateState((prevState) => ({ ...prevState, ...newState }));
+                    }}
+                    initialValues={state}
+                >
+                    <Field
+                        label={<FormattedMessage id="comet.blocks.link.phone" defaultMessage="Phone Number" />}
+                        name="phone"
+                        component={FinalFormInput}
+                        fullWidth
+                        validate={(phone: string) => {
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/blocks/PhoneLinkBlock.tsx:38]: Please use validatePhoneNumber here

File: packages/admin/cms-admin/src/blocks/PhoneLinkBlock.tsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+import * as React from "react";
+import { FormattedMessage } from "react-intl";
+
+import { PhoneLinkBlockData, PhoneLinkBlockInput } from "../blocks.generated";
+import { isPhoneNumber } from "../validation/isPhoneNumber";
+
+export const PhoneLinkBlock: BlockInterface<PhoneLinkBlockData, PhoneLinkBlockData, PhoneLinkBlockInput> = {
+    ...createBlockSkeleton(),
+
+    name: "Phone",
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/blocks/PhoneLinkBlock.tsx:12]:
```suggestion
    name: "PhoneLink",
```

File: packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import { isEmail } from "class-validator";
+import * as React from "react";
+import { FormattedMessage } from "react-intl";
+
+import { EmailLinkBlockData, EmailLinkBlockInput } from "../blocks.generated";
+
+export const EmailLinkBlock: BlockInterface<EmailLinkBlockData, EmailLinkBlockData, EmailLinkBlockInput> = {
+    ...createBlockSkeleton(),
+
+    name: "Email",
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx:12]:
```suggestion
    name: "EmailLink",
```

File: packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        return state.email ? isEmail(state.email) : true;
+    },
+
+    AdminComponent: ({ state, updateState }) => {
+        return (
+            <SelectPreviewComponent>
+                <BlocksFinalForm
+                    onSubmit={(newState: EmailLinkBlockData) => {
+                        updateState((prevState) => ({ ...prevState, ...newState }));
+                    }}
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/blocks/EmailLinkBlock.tsx:30]:
```suggestion
                    onSubmit={updateState}
```

**Comment by johnnyomair** [comment]
Why are the blocks in `@comet/blocks-api` but in `@comet/cms-admin`, and not `@comet/blocks-admin`?

**Comment by SebiVPS** [comment]
> Why are the blocks in `@comet/blocks-api` but in `@comet/cms-admin`, and not `@comet/blocks-admin`?

@johnnyomair i put it there because there are more other blocks around in `@comet/cms-admin`. In `@comet/blocks-admin` there is only a deprecated `SpaceBlock`and a `YoutTubeBlock`.
We can move it to the other package if you want. I don't know where it really belongs.

**Comment by johnnyomair** [comment]
> > Why are the blocks in `@comet/blocks-api` but in `@comet/cms-admin`, and not `@comet/blocks-admin`?
> 
> @johnnyomair i put it there because there are more other blocks around in `@comet/cms-admin`. In `@comet/blocks-admin` there is only a deprecated `SpaceBlock`and a `YoutTubeBlock`. We can move it to the other package if you want. I don't know where it really belongs.

I'd say move it `@comet/cms-api`. We plan to dissolve `@comet/blocks-api` and `@comet/blocks-admin` into the respective cms packages.

File: packages/api/cms-api/src/index.ts [hunk 13]
```diff
@@ -59,6 +61,7 @@ export { PaginatedResponseFactory } from "./common/pagination/paginated-response
 export { SortArgs } from "./common/sorting/sort.args";
 export { SortDirection } from "./common/sorting/sort-direction.enum";
 export { IsNullable } from "./common/validators/is-nullable";
+export { IsPhoneNumber, isPhoneNumber, IsPhoneNumberConstraint } from "./common/validators/is-phone-number";
```
Comment (johnnyomair) [inline packages/api/cms-api/src/index.ts:64]:
```suggestion
export { IsPhoneNumber, isPhoneNumber } from "./common/validators/is-phone-number";
```

I don't think it's necessary to export the constraint. Should we even export the validator? If yes, a changeset would be nice.

File: packages/api/cms-api/src/common/validators/is-phone-number.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+            target: object.constructor,
+            propertyName,
+            validator: IsPhoneNumberConstraint,
+        });
+    };
+};
+
+@ValidatorConstraint({ name: "IsPhoneNumber" })
+@Injectable()
+export class IsPhoneNumberConstraint implements ValidatorConstraintInterface {
```
Comment (johnnyomair) [inline packages/api/cms-api/src/common/validators/is-phone-number.ts:23]:
```suggestion
class IsPhoneNumberConstraint implements ValidatorConstraintInterface {
```

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**